### PR TITLE
Glob match migrations

### PR DESCRIPTION
**Pull request by iilei**
Following up on the discussion at #2756

### CODE REVIEW

**Comment by iilei** [comment]
was filing this one by mistake 😅 - will have another session fixing the build

**Comment by iilei** [comment]
Changelog Gist:

__Breaking Changes:__
- Options `directory` and `loadExtensions` superseded by glob pattern matching options `globPatterns` handed over to [globby](https://github.com/sindresorhus/globby)

__New features:__
- Option [`globOptions.gitignore=false`](https://github.com/sindresorhus/globby#gitignore) available for configuration of globbing behaviour.

**Comment by iilei** [comment]
The failing tests say `mergedConfig.migrationSource.getMigrationDirs` is not implemented. Is that probably related to some caching of compiled sources?

File: src/migrate/Migrator.js [hunk 1]
```diff
@@ -403,11 +405,17 @@ export function getMergedConfig(config, currentConfig) {
...
-      mergedConfig.directory,
-      mergedConfig.sortDirsSeparately
+      mergedConfig.globPatterns,
+      mergedConfig.sortDirsSeparately,
+      mergedConfig.globOptions
     );
   }
 
+  if (!mergedConfig.directory) {
+    const directory = await mergedConfig.migrationSource.getMigrationDirs();
```
Comment (JakeGinnivan) [inline src/migrate/Migrator.js]: This is another breaking change. What would I return for say the `WebpackMigrationSource` which doesn't have directories?

File: src/migrate/Migrator.js [hunk 2]
```diff
@@ -32,11 +35,10 @@ function LockError(msg) {
 inherits(LockError, Error);
 
 const CONFIG_DEFAULT = Object.freeze({
-  extension: 'js',
-  loadExtensions: DEFAULT_LOAD_EXTENSIONS,
+  globPatterns: DEFAULT_GLOB_PATTERNS,
```
Comment (JakeGinnivan) [inline src/migrate/Migrator.js]: This could also be a non-breaking change. If globPatterns are not specified then the directory is used to build the glob?

Comment (iilei) [reply src/migrate/Migrator.js]: I thought of that possibility but there is [already a breaking change related to the use case addressed](#2756 ) with this patch. Additionally I thought it was better this way in favor of a more concise API.

**Comment by iilei** [comment]
I am pausing the work on this one for now. If someone likes to take over, I'd appreciate it

**Pull request closed without merging**